src/nvim/os/job.c

```diff
@@ -55,6 +56,22 @@ uv_signal_t schld;
 #endif
 // Callbacks for libuv
 
+static int count = 0;  // A cyclical counter used for assigning job IDs.
+
+/// Returns a new job ID within the range `[1, INT_MAX]`.
+///
+/// This number will be unique between all jobs that have not been stopped, but
+/// cycle back to one after reaching `INT_MAX`. This makes it highly unlikely
+/// for two jobs to have the same ID.
+static int new_job_id(void)
+{
+  count = count < INT_MAX ? count + 1 : 1;
```

Comment:
By the way, now nobody checks for the condition “job table is full”: this is now equivalent to an infinite cycle.